### ISSUE TITLE
Avoid R_PreserveObject for performance reasons

### ISSUE THREAD

**Issue by clauswilke** [extendr/extendr#27]
Just filing this here as an issue that should be addressed at some point. The current code uses R_PreserveObject to create objects that are "owned" (not really) by Rust and protected from the GC. See e.g. here:
https://github.com/extendr/extendr/blob/5245703e99bcd8b65d647149f954c680b9966fcc/extendr-api/src/robj.rs#L1057

This works fine as proof of concept but should be replaced eventually. It will not scale. (Until recently, it could crash R.)

See a detailed explanation of this issue here: https://github.com/RcppCore/Rcpp/issues/1081
See the code that cpp11 wrote to work around this issue here:
https://github.com/r-lib/cpp11/blob/f1f971ffdd94d7949b7881fccadef825ba17a0a4/inst/include/cpp11/protect.hpp#L209-L364

Basically, we need to build our own data structure (e.g., an R linked list) that is understood by R's GC and that can rapidly created and remove protected objects.

**Comment by andy-thomason** [comment]
Yes.

I had planned to do this.

They have improved performance of R_PreserveObject in  later releases
but it also suffers the danger that the objects are not reference counted
and may be double-freed.

Having a hash table of indices and a single preserved R list would
root the garbage collector for any objects we want to keep.

In the meantime, I was hoping to avoid generating any R objects
except for returns by providing Rust alternatives.

Bad things would be vectors of Robj which would make many, many
unnecessary gc roots.

Also, vectors and matrices of SEXP are possible in R, which would be hard to map.

Happy to accept a PR which does this at any time.

**Comment by clauswilke** [comment]
I think the cpp11 approach is both the most performant and most consistent with how R thinks about object ownership: create an R double-linked list and for owned objects store a pointer to an element in the list, which then points to the actual object. It's one additional dereferencing that is required to access objects, but adding/removing elements is trivial.

This is not high on my priority list right now, because the current approach works, but eventually this will become a bottleneck, even if the feature is used sparingly. In the case that prompted the Rcpp issue I linked, I was creating a handful of preserved objects per ggplot2 plot generated. Should not be a big issue, but became one when somebody tried to generate several thousand ggplot2 plots within the same R session.
